### PR TITLE
fix(web): build config cleanup, Safari & mobile-device bugfixes (thanks to CI checks) 🧩

### PR DESCRIPTION
**Pull request by jahorton**
The primary, overarching goal of this PR:  to get Web passing its CI checks again, even with the new unit tests included.  It's been 'failing' for a long time on this 🧩 feature branch, and we can finally fix that.

Secondary goal:  adhering to our goals for build-script organization & design, along with the wish to keep them reasonably optimized.

### Aspect 1 - build script/config cleanup

This focuses on two specific details:

1. Updates the feature-branch build scripts to utilize #8742.
2. As noted on the base branch, I'd gone a bit too quick-and-dirty with the tsconfigs.  The "reference" entries needed some cleaning up.

### Aspect 2 - CI test configuration

It's been noted before that each build script should only test code within its own domain.  Any larger-scale, cross-project testing should be handled via CI script.  Up until now... Web kind of ignored this.

Well, that's no longer.  Tests of `@keymanapp/keyboard-processor` and `@keymanapp/input-processor` are no longer triggered by web's `./build.sh` or `./test.sh`; instead, they're now triggered via `./ci.sh test`.

### Aspect 3 - a minor optimization:  untargeted action checks

The short version:  I've added one small tweak to `builder_parse`'s parsing procedures.  When enumerating all possible `target` entries for an `action` specified without a target, it now adds one bonus entry:  `action:project`.  As `:project` has been noted to be the default target if none is specified, I felt it the natural fit to denote an _**explicitly project-wide, untargeted**_ action.

```bash
if builder_has_action build; then
  # will trigger from `build`, `build:child`, or `build:target`
fi

if builder_has_action build:project; then
  # will only trigger from `build`; it will not trigger for `build:child` or `build:target`
fi
```

More details at the `master` 🍒-pick version: #8831.

I _can_ undo it here without too much issue if necessary.

### Aspect 4 - bugfix for Safari-specific behaviors

While #8826 _mostly_ gets Web's unit tests back online, I ran into some odd bugs coming from tests run on Safari.  Long story short, WebKit has intended selection + focus interactions that would appear as bugs on other browsers, and failure to account for this was causing issues.

Some light reading:  https://bugs.webkit.org/show_bug.cgi?id=224425

Fortunately, it's very simple to prevent this from causing an issue now that we're aware of it.

While I'd normally make a separate issue for this...

1. It's extremely orthogonal to the other changes and is very focused - it's pretty cheap to "throw in" here.
2. Including it in this build, along with the other changes, should validate "Aspect 3" below and finally give us passing Web CI test-builds again.  The other aspects are the reason this KMW bug was discovered.

### Aspect 5 - mobile-device bugfixes

This is what happens when you don't run the touch-device test suite locally first and rely on CI to handle it.  I probably could have streamlined this somewhat.

So, fun fact: the JS engine can't detect infinite recursion when MutationObservers are involved - that is, if an observer's handler re-triggers the observer itself.  That was causing execution timeouts, as this behavior was occurring on the Android test target and locking up the integration test suite.  The underlying cause:  the observer was being turned _back on_ during engine-shutdown due to an oversight, leaving an old engine's observer active while a new engine was loaded.  Yuck.

Finally, it adds better zero-stub handling for default-keyboard behavior on touch devices; that case was missing a proper null guard.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
... and the Safari test timed out this time - the old "BrowserStack didn't connect" problem.  Fortunately, I can say two things:

- all tests that ran did pass
- it did pass with an earlier test run: https://build.palaso.org/buildConfiguration/Keymanweb_TestPullRequests/383676?buildTab=tests&name=Safari

So... I'll need to try running it again later, but I think this indicates that I've (finally) patched everything up properly. 🤞

**Comment by jahorton** [comment]
As of https://github.com/keymanapp/keyman/pull/8830/commits/0e422271b7f06d8f3ae7fd1a86e4e24a7b4cf8ff, it passed all its unit & integrated tests!  Alas, I had neglected to dry-run the CI artifact upload stages...

> stat: cannot stat '/c/BuildAgent/work/99b311828f4ee7c/keyman/web/build/app/web/release/keymanweb.js': No such file or directory

`/build/app/web/release` => `/build/app/browser/release` now.  And so the file-size check validation fell over and crashed the run.  🤦

**Comment by jahorton** [comment]
🎉 As of commit https://github.com/keymanapp/keyman/pull/8830/commits/7960c3cb43871d1c30248a6b80bf681e096e7d94, 🎉 

image

We now have the first passing ES-module based Web build.

File: resources/build/test/test.sh [hunk 1]
```diff
@@ -113,6 +113,23 @@ builder_describe \
   "--zoom,-z    Use zoom mode" \
   "--feature=FOO Enable feature foo"
 
+#----------------------------------------------------------------------
+# Test implicit :project targets
+
```
Comment (mcdurdin) [inline resources/build/test/test.sh:118]: If we can move the `:project` stuff into #8831 that would help :grin:

**Review by mcdurdin — APPROVED** [review]

**Pull request merged**